relayer/relays/beefy/on-demand-sync.go

```diff
@@ -72,101 +83,93 @@ func (relay *OnDemandRelay) Start(ctx context.Context) error {
 	}
 
 	gatewayAddress := common.HexToAddress(relay.config.Sink.Contracts.Gateway)
-	gatewayContract, err := contracts.NewGateway(gatewayAddress, relay.ethereumConn.Client())
+	gatewayContractV2, err := contracts.NewGateway(gatewayAddress, relay.ethereumConn.Client())
+	if err != nil {
+		return fmt.Errorf("create gateway client: %w", err)
+	}
+	relay.gatewayContractV2 = gatewayContractV2
+
+	gatewayContractV1, err := contractV1.NewGateway(gatewayAddress, relay.ethereumConn.Client())
 	if err != nil {
 		return fmt.Errorf("create gateway client: %w", err)
 	}
-	relay.gatewayContract = gatewayContract
+	relay.gatewayContractV1 = gatewayContractV1
 
 	relay.tokenBucket.Start(ctx)
 
 	for {
-		sleep(ctx, time.Minute*1)
-		log.Info("Starting check")
-
-		paraNonce, ethNonce, err := relay.queryNonces(ctx)
-		if err != nil {
-			if errors.Is(err, context.Canceled) {
-				return nil
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(time.Second * 30):
+			log.Info("Starting check")
+			err = relay.syncV1(ctx)
+			if err != nil {
+				return fmt.Errorf("sync for v1 message: %w", err)
+			}
+			err = relay.syncV2(ctx)
+			if err != nil {
+				return fmt.Errorf("sync for v2 message: %w", err)
 			}
-			log.WithError(err).Error("Query nonces")
-			continue
 		}
+	}
+}
 
-		log.WithFields(log.Fields{
-			"paraNonce": paraNonce,
-			"ethNonce":  ethNonce,
-		}).Info("Nonces checked")
+func (relay *OnDemandRelay) syncV1(ctx context.Context) error {
+	paraNonce, ethNonce, err := relay.queryNonces(ctx)
+	if err != nil {
+		return fmt.Errorf("Query nonces: %w", err)
+	}
 
-		if paraNonce > ethNonce {
+	log.WithFields(log.Fields{
+		"paraNonce": paraNonce,
+		"ethNonce":  ethNonce,
+	}).Info("Nonces checked")
 
-			// Check if we are rate-limited
-			if !relay.tokenBucket.TryConsume(1) {
-				log.Info("Rate-limit exceeded")
-				continue
-			}
+	if paraNonce > ethNonce {
 
-			log.Info("Performing sync")
+		// Check if we are rate-limited
+		if !relay.tokenBucket.TryConsume(1) {
+			return fmt.Errorf("Rate-limit exceeded")
+		}
 
-			beefyBlockHash, err := relay.relaychainConn.API().RPC.Beefy.GetFinalizedHead()
-			if err != nil {
-				if errors.Is(err, context.Canceled) {
-					return nil
-				}
```

Comment:
Since we've moved the relay into an external goroutine,  checking for context cancellation here is not nessessary.

Reply:
fyi, The purpose of this check is to exit the process cleanly when I press 'Ctrl-C'. I've setup the relayers to handle these these signals using context cancellation.

So you got to make sure to test that flow.

Reply:
Yeah, I've tested the exit flow - it responds to `Ctrl-C` correctly.